python/katana/local/_graph.pyx

```diff
@@ -275,6 +275,14 @@ cdef class GraphBase:
         """
         return str(self.underlying_property_graph().rdg_dir(), encoding="UTF-8")
 
+    @property
+    def node_type_names(self):
+        """
+        The names of atomic node types in the graph.
+
+        :rtype: list
+        """
+        return self.underlying_property_graph().GetNodeTypeManager().GetAtomicTypeNames()
```

Comment:
For now the object can just provide the name. But I think we will want it to do more eventually. It could be declared as:

```
cdef class EntityType:
  def __init__(self, type_manager, type_id):
    self._type_manager = type_manager
    self._type_id = type_id
  def __str__(self):
    return self._type_manager.GetAtomicTypeName(self.type_id)
```
I didn't test this, but it should be close to what you will want. You can then instantiate them based on a vector of IDs with a list comprehension.

Reply:
I see your point, but I would like to balance it with a concern for maintaining a super simple python interface. Data scientists are sometimes not the best coders and I worry that adding too many abstractions would scare some of them away. I am basing my implementation roughly on DGL, which uses strings for type names in their HeteroGraph.

What if we implemented this EntityType class but still exported an API that lets users just deal in strings if that's what they're comfortable with? I would keep the node_type_names property but implement it to use the EntityType class.

Reply:
I see your point, but I think we should push for better APIs to help data scientists improve. I think we can make the EntityType actually pretend to be a string in most cases, but still carry the extra information and have the extra properties. This might give us a best of both worlds.

Also I expected `node_type_names` or `node_types` (whatever we want to call it) to return EntityTypes. I'm not sure what else you were envisioning.

I would prefer we start with a strict API with a EntityType and then relax it as needed. It's much easier to add support for other types than remove support for unsafe code that may already have been written. One reason for EntityType, BTW, is that it could check whether the type is of the correct kind, node vs edge, when the type is used. Which would be quite useful for catching mistakes.

All this said, I can come back with a PR to make the API a bit stricter after you merge this if you don't think it's easy to do in this PR.

Reply:
Seems easy enough, and you make a compelling case. I'm working on it. I do just want to keep this concern for the lowest common denominator python user in mind though as we develop the API. For now, EntityTypes it is.

Reply:
You are right and I do have that in mind: https://katanagraph.atlassian.net/browse/KAT-628. It should be an epic. There are lots of bits involved. I'll change it. Feel free to add things to it if you think there are more potential confusions to be avoided.

My overall goal is to provide an API that is mostly familiar to a data-scientist, but also provides a bit more safety than they may be useful. Things like raising an exception if they try to index a node property with an edge index. But I still want it to integrate well with other code, so it will not be "too safe" since any special index types will still need to support conversion to int for indexing numpy arrays and the like. I'd love your input on all this.

Reply:
I would welcome a meeting about Types and type management.  I like Arthur's idea of EntityTypes as an actual type that might have a string or a numeric representation.  If I got those from the Cypher engine, I'd be happy to process them and pass them to the partitioner.  Of course the C++ and python types don't need to be exactly 1:1, but the way Aurthur describes it, this is an opportunity to use the type system to our advantage for little performance cost.
What I don't like about the current interface is how clients have to understand the underlying representation to do things like iterate over all compound types.  Ranges are fine if our ID space is always guaranteed to be dense.  Is it?